File: elixir/lib/datagrout_conduit/deliverables.ex
defmodule DatagroutConduit.Deliverables do
  @moduledoc "Work product registration, listing, and retrieval."

  @doc "Register a work product (`data-grout/deliverables.register`)."
  @spec register(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def register(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "deliverables.register", opts)
  end

  @doc "List deliverables with optional semantic search (`data-grout/deliverables.list`)."
  @spec list(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def list(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "deliverables.list", opts)
  end

  @doc "Get a specific deliverable by reference (`data-grout/deliverables.get`)."
  @spec get(GenServer.server(), String.t()) :: {:ok, map()} | {:error, term()}
  def get(client, ref_id) do
    DatagroutConduit.Client.dg(client, "deliverables.get", %{"ref" => ref_id})
  end
end

File: python/src/datagrout/conduit/namespaces/deliverables.py
"""Deliverables namespace — work product tracking and retrieval."""

from typing import Any, Dict, Optional


class DeliverablesNamespace:
    """Work product registration, listing, and retrieval."""

    def __init__(self, client: Any):
        self._client = client

    async def register(self, params: Dict[str, Any]) -> Any:
        """Register a work product (`data-grout/deliverables.register`)."""
        self._client._warn_if_not_dg("deliverables.register")
        self._client._ensure_initialized()
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/deliverables.register", params)
        )

    async def list(self, params: Optional[Dict[str, Any]] = None) -> Any:
        """List deliverables with optional semantic search (`data-grout/deliverables.list`)."""
        self._client._warn_if_not_dg("deliverables.list")
        self._client._ensure_initialized()
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/deliverables.list", params or {})
        )

    async def get(self, ref_id: str) -> Any:
        """Get a specific deliverable by reference (`data-grout/deliverables.get`)."""
        self._client._warn_if_not_dg("deliverables.get")
        self._client._ensure_initialized()
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/deliverables.get", {"ref": ref_id})
        )
